src/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/Formatters/Binary/BinaryFormatter.cs

```diff
@@ -0,0 +1,119 @@
+        {
+            if (serializationStream == null)
+            {
+                throw new ArgumentNullException(nameof(serializationStream), SR.Format(SR.ArgumentNull_WithParamName, serializationStream));
+            }
+            if (serializationStream.CanSeek && (serializationStream.Length == 0))
+            {
+                throw new SerializationException(SR.Serialization_Stream);
+            }
+
+            var formatterEnums = new InternalFE()
+            {
+                _FEtypeFormat = _typeFormat,
+                _FEserializerTypeEnum = InternalSerializerTypeE.Binary,
+                _FEassemblyFormat = _assemblyFormat,
+                _FEsecurityLevel = _securityLevel,
+            };
+
+            var reader = new ObjectReader(serializationStream, _surrogates, _context, formatterEnums, _binder)
+            {
+                _crossAppDomainArray = _crossAppDomainArray
+            };
+            var parser = new BinaryParser(serializationStream, reader);
+            return reader.Deserialize(handler, parser, check);
+        }
+
+        public object Deserialize(Stream serializationStream, HeaderHandler handler) => 
+            Deserialize(serializationStream, handler, check: true);
+
+        [ComVisible(false)]
+        public object UnsafeDeserialize(Stream serializationStream, HeaderHandler handler) => 
+            Deserialize(serializationStream, handler, check: false);
+
+        public void Serialize(Stream serializationStream, object graph) => 
+            Serialize(serializationStream, graph, headers: null);
+
+        public void Serialize(Stream serializationStream, object graph, Header[] headers) => 
+            Serialize(serializationStream, graph, headers, check: true);
+
+        internal void Serialize(Stream serializationStream, object graph, Header[] headers, bool check)
+        {
+            if (serializationStream == null)
+            {
+                throw new ArgumentNullException(nameof(serializationStream), SR.Format(SR.ArgumentNull_WithParamName, serializationStream));
+            }
+
+            var formatterEnums = new InternalFE()
+            {
+                _FEtypeFormat = _typeFormat,
+                _FEserializerTypeEnum = InternalSerializerTypeE.Binary,
+                _FEassemblyFormat = _assemblyFormat,
+            };
+
+            var sow = new ObjectWriter(_surrogates, _context, formatterEnums, _binder);
+            BinaryFormatterWriter binaryWriter = new BinaryFormatterWriter(serializationStream, sow, _typeFormat);
+            sow.Serialize(graph, headers, binaryWriter, check);
+            _crossAppDomainArray = sow._crossAppDomainArray;
+        }
+
+        internal static TypeInformation GetTypeInformation(Type type)
+        {
+            lock (s_typeNameCache)
+            {
+                TypeInformation typeInformation;
+                if (!s_typeNameCache.TryGetValue(type, out typeInformation))
```

Comment:
Dictionary is generally preferable over Hashtable, but Hashtable has one critical behavior/guarantee that Dictionary does not: it's safe for readers to use the Hashtable concurrently with a single writer (it's not safe for multiple writers to be writing concurrently).  Some of the uses of Hashtable in the desktop codebase were taking advantage of that, with a pattern like:
```C#
if (hashtable.ContainsKey(key)) return hashtable[key];
lock (syncobj)
{
    if (hashtable.ContainsKey(key)) return hashtable[key];
    hashtable.Add(key, ProduceValue(key));
}
```
I converted those to use ConcurrentDictionary, since the original developer apparently felt it was important to make the reads lock-free.  In other cases, like this one, all accesses were protected by the lock, so I just made those be Dictionary, as I didn't have any perf data to suggest that a ConcurrentDictionary would be beneficial.  If we find it's a bottleneck, we can always replace it later.